### PR TITLE
[BEAM-1938] Update Signature of PTransformOverrideFactory

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
This enables replacements to be reobtained with the entire transform
that is being replaced.

This is required when Side Inputs are part of the input of the
PTransform Application, as PTransforms are not applied to their side
inputs.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @jkff

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/WriteWithShardingFactoryTest.java [hunk 1]
```diff
@@ -195,20 +207,13 @@ public void keyBasedOnCountFnManyElementsExtraShards() throws Exception {
...
-  }
-
   private static class TestSink extends Sink<Object> {
     @Override
     public void validate(PipelineOptions options) {}
 
     @Override
     public WriteOperation<Object, ?> createWriteOperation(PipelineOptions options) {
-      throw new IllegalArgumentException("Should not be used");
+      return null;
```
Comment (jkff) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/WriteWithShardingFactoryTest.java]: Why this change?

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java [hunk 2]
```diff
@@ -1336,15 +1347,17 @@ private StreamingPubsubIOWriteOverrideFactory(DataflowRunner runner) {
       this.runner = runner;
     }
 
-    @Override
-    public PTransform<PCollection<T>, PDone> getReplacementTransform(
-        PubsubUnboundedSink<T> transform) {
-      return new StreamingPubsubIOWrite<>(runner, transform);
+
+    public PCollection<T> getInput(Map<TupleTag<?>, PValue> inputs) {
```
Comment (jkff) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java]: This is not necessary anymore, right?

File: runners/apex/src/main/java/org/apache/beam/runners/apex/ApexRunner.java [hunk 3]
```diff
@@ -258,9 +260,15 @@ protected String getKindString() {
...
-      public PTransform<PCollection<InputT>, PCollectionView<OutputT>> getReplacementTransform(
-          GloballyAsSingletonView<InputT, OutputT> transform) {
-        return new StreamingCombineGloballyAsSingletonView<>(transform);
+      public PTransformReplacement<PCollection<InputT>, PCollectionView<OutputT>>
+          getReplacementTransform(
+              AppliedPTransform<
+                      PCollection<InputT>, PCollectionView<OutputT>,
+                      GloballyAsSingletonView<InputT, OutputT>>
+                  transform) {
+        return PTransformReplacement.of(
```
Comment (tgroh) [inline runners/apex/src/main/java/org/apache/beam/runners/apex/ApexRunner.java]: There's definitely utilities we can add. Due to the reason this PR exists, they need to handle an iterable that contains more than a single input, where only one is the "main input", so I've migrated to that everywhere.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoMultiOverrideFactory.java [hunk 4]
```diff
@@ -62,8 +62,19 @@
     implements PTransformOverrideFactory<
         PCollection<? extends InputT>, PCollectionTuple, MultiOutput<InputT, OutputT>> {
   @Override
+  public PTransformReplacement<PCollection<? extends InputT>, PCollectionTuple>
+      getReplacementTransform(
+          AppliedPTransform<
+                  PCollection<? extends InputT>, PCollectionTuple, MultiOutput<InputT, OutputT>>
+              transform) {
+    PCollection<InputT> singletonMainInput =
```
Comment (jkff) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoMultiOverrideFactory.java]: inline variable perhaps, for consistency with all other cases

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/DeduplicatedFlattenFactory.java [hunk 5]
```diff
@@ -47,41 +49,19 @@
...
-            output = output.and(duplicated);
-          }
-        }
-        return output.apply(Flatten.<T>pCollections());
-      }
-    };
+  public PTransformReplacement<PCollectionList<T>, PCollection<T>> getReplacementTransform(
+      AppliedPTransform<PCollectionList<T>, PCollection<T>, PCollections<T>> transform) {
+    return PTransformReplacement.of(
+        getInput(transform.getInputs(), transform.getPipeline()),
```
Comment (jkff) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/DeduplicatedFlattenFactory.java]: Note: this is one of only two cases that isn't PBegin or getSingletonMainInput.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/EmptyFlattenAsCreateFactory.java [hunk 6]
```diff
@@ -49,20 +50,15 @@
...
     checkArgument(
-        inputs.isEmpty(),
+        transform.getInputs().isEmpty(),
         "Unexpected nonempty input %s for %s",
-        inputs,
+        transform.getInputs(),
         getClass().getSimpleName());
-    return PCollectionList.empty(p);
+    return PTransformReplacement.of(
+        PCollectionList.<T>empty(transform.getPipeline()), new CreateEmptyFromList<T>());
```
Comment (jkff) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/EmptyFlattenAsCreateFactory.java]: I wonder if you can change this to be PBegin too?

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PTransformReplacements.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+import org.apache.beam.sdk.values.TupleTag;
+
+/**
+ */
+public class PTransformReplacements {
+  /**
+   * Gets the singleton input of an {@link AppliedPTransform}, ignoring any additional inputs
+   * returned by {@link PTransform#getAdditionalInputs()}.
+   */
+  public static <T> PCollection<T> getSingletonMainInput(
```
Comment (jkff) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PTransformReplacements.java]: All calls but one use either this or PBegin - you could have a utility method that handles both, since I guess the need for PBegin can be detected automatically by seeing that the transform has no inputs at all.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoMultiOverrideFactory.java (hunk 4 above)
Comment (tgroh) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoMultiOverrideFactory.java]: Done.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/EmptyFlattenAsCreateFactory.java (hunk 6 above)
Comment (tgroh) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/EmptyFlattenAsCreateFactory.java]: This would work, but the types don't actually line up anymore, so I'd prefer not to do it, especially given that this was broken for an extended period precisely because the types didn't line up.

**Comment by tgroh** [comment]
I'm going to merge on green and file a follow up to remove some duplication from these classes.

**Comment by tgroh** [comment]
https://issues.apache.org/jira/browse/BEAM-1982 filed
